File: web-app/app/javascript/controllers/user_list_state_controller.js
import { Controller } from "@hotwired/stimulus"

// Bump when the persisted state shape changes (e.g. memberships went from
// [list_id] to [{list_id, item_id}] tuples). On load, caches stamped with a
// different schema are discarded so users don't render against stale shapes.
const STATE_SCHEMA = 2

// Singleton controller. Hydrates per-user list/membership state from
// localStorage, then refreshes from /user_list_state. Broadcasts events on
// window so per-card widgets and the singleton modal can re-render.
//
// Storage shape: { version, domain, lists, memberships }
// Event names emitted on window:
//   - "user-list-state:loaded"  (cache hit, may be stale)
//   - "user-list-state:updated" (network response superseded cache)
//   - "user-list-state:cleared" (sign-out detected)
export default class extends Controller {
  static values = {
    url: { type: String, default: "/user_list_state" }
  }

  connect() {
    this.domain = document.body.dataset.domain || "books"
    this.storageKey = `tg:user_list_state:${this.domain}`
    this.signedIn = false
    this.cache = null
    // CSRF token is held in memory only — never persisted to localStorage.
    // The cached HTML's <meta name="csrf-token"> is unreliable on CDN-cached
    // pages, so mutations must use the token from the latest /user_list_state.
    this.csrf = null
    this._inflightRefresh = null

    this._onAuthSuccess = this._onAuthSuccess.bind(this)
    this._onAuthSignout = this._onAuthSignout.bind(this)
    window.addEventListener("auth:success", this._onAuthSuccess)
    window.addEventListener("auth:signout", this._onAuthSignout)

    // Hydration is gated on the tg_uid cookie (set by AuthController on sign-in,
    // backfilled by /user_list_state). This prevents a previous user's cache
    // from rendering on a shared browser after sign-out / session expiry.
    if (this.cookieUid()) {
      this._hydrateFromStorage()
      this.refresh()
    } else {
      // No signed-in marker — discard any stale cache and stay anonymous.
      // We deliberately skip the network fetch: the endpoint requires auth
      // and would just return 401.
      this._clearStorage()
    }
  }

  // Reads the non-HttpOnly tg_uid cookie set by AuthController on sign-in.
  // Returns the user id as a string, or null if the cookie is absent.
  cookieUid() {
    const m = document.cookie.match(/(?:^|;\s*)tg_uid=([^;]+)/)
    return m ? decodeURIComponent(m[1]) : null
  }

  disconnect() {
    window.removeEventListener("auth:success", this._onAuthSuccess)
    window.removeEventListener("auth:signout", this._onAuthSignout)
  }

  // Returns the current in-memory state (or null if none).
  state() {
    return this.cache
  }

  isSignedIn() {
    return this.signedIn
  }

  // Returns the in-memory CSRF token (only available after a successful fetch).
  csrfToken() {
    return this.csrf
  }

  // Ensures a fresh CSRF token is available before a mutation, fetching once
  // if none. Coalesces concurrent calls onto a single in-flight request.
  async ensureCsrf() {
    if (this.csrf) return this.csrf
    await this.refresh()
    return this.csrf
  }

  // Re-fetch /user_list_state. If the response version is newer than the cached
  // version, persist + dispatch updated. On 401, treat as signed out.
  // Concurrent callers share a single in-flight promise.
  refresh() {
    if (this._inflightRefresh) return this._inflightRefresh
    this._inflightRefresh = this._doRefresh().finally(() => {
      this._inflightRefresh = null
    })
    return this._inflightRefresh
  }

  async _doRefresh() {
    let response
    try {
      response = await fetch(this.urlValue, {
        headers: { Accept: "application/json" },
        credentials: "same-origin"
      })
    } catch (err) {
      console.warn("user-list-state: network error", err)
      return
    }

    if (response.status === 401) {
      this._clearStorage()
      this._clearCookieUid()
      this.signedIn = false
      this.cache = null
      this.csrf = null
      this._dispatch("user-list-state:cleared")
      return
    }

    if (!response.ok) {
      console.warn("user-list-state: unexpected status", response.status)
      return
    }

    const data = await response.json()
    this.signedIn = true
    this.csrf = data.csrf_token || null
    // Strip csrf_token from the persisted shape so it never lands in localStorage.
    const { csrf_token: _csrf, ...persistable } = data

    // The network response is authoritative — always replace the cache.
    // (We previously skipped the write when versions matched, but local
    // optimistic mutations stamp the cache with Date.now(), which can be
    // ahead of server `user.updated_at.to_i`, suppressing legitimate updates.)
    this.cache = persistable
    this._writeStorage(persistable)
    this._dispatch("user-list-state:updated", { state: persistable })
  }

  // Optimistically apply a mutation result and rebroadcast. Called by the modal
  // after a successful add/remove/create. `memberships` is the new full array
  // of {list_id, item_id} tuples for this (listableType, listableId).
  applyMutation({ listableType, listableId, memberships, addedList, removedListId }) {
    if (!this.cache) return
    const next = JSON.parse(JSON.stringify(this.cache))

    if (addedList) {
      const exists = next.lists.find((l) => l.id === addedList.id)
      if (!exists) next.lists.push(addedList)
    }

    if (removedListId) {
      next.lists = next.lists.filter((l) => l.id !== removedListId)
    }

    if (listableType && listableId) {
      const key = String(listableId)
      next.memberships[listableType] ||= {}
      next.memberships[listableType][key] = memberships
      if (!memberships.length) delete next.memberships[listableType][key]
    }

    next.version = Math.floor(Date.now() / 1000)
    this.cache = next
    this._writeStorage(next)
    this._dispatch("user-list-state:updated", { state: next })
  }

  _hydrateFromStorage() {
    try {
      const raw = window.localStorage.getItem(this.storageKey)
      if (!raw) return
      const parsed = JSON.parse(raw)
      if (parsed?._schema !== STATE_SCHEMA) {
        // Stored under an older shape — discard so we don't render bad UI.
        window.localStorage.removeItem(this.storageKey)
        return
      }
      // Bind cache to the current signed-in user. If cookie uid doesn't match
      // the cached user_id, this is a different user on a shared browser —
      // discard the previous user's data instead of rendering it.
      const cookieUid = this.cookieUid()
      if (parsed.user_id != null && String(parsed.user_id) !== cookieUid) {
        window.localStorage.removeItem(this.storageKey)
        return
      }
      this.cache = parsed
      this.signedIn = true
      this._dispatch("user-list-state:loaded", { state: this.cache })
    } catch (err) {
      console.warn("user-list-state: storage read failed", err)
    }
  }

  _writeStorage(state) {
    try {
      window.localStorage.setItem(this.storageKey, JSON.stringify({...state, _schema: STATE_SCHEMA}))
    } catch (err) {
      // QuotaExceededError or similar — fall back to in-memory only.
      console.warn("user-list-state: storage write failed (quota?)", err)
    }
  }

  _clearStorage() {
    try { window.localStorage.removeItem(this.storageKey) } catch (_) { /* ignore */ }
  }

  // Clears the non-HttpOnly tg_uid cookie. Used when the server reports the
  // session is no longer valid (401) so a stale cookie doesn't keep gating
  // future hydration as if signed in.
  _clearCookieUid() {
    document.cookie = "tg_uid=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/; SameSite=Lax"
  }

  _dispatch(name, detail = {}) {
    window.dispatchEvent(new CustomEvent(name, { detail }))
  }

  _onAuthSuccess() {
    this.refresh()
  }

  _onAuthSignout() {
    this._clearStorage()
    this.signedIn = false
    this.cache = null
    this.csrf = null
    this._dispatch("user-list-state:cleared")
  }
}

File: web-app/app/controllers/user_list_state_controller.rb
class UserListStateController < ApplicationController
  include Cacheable
  include JsonErrorResponses

  # State endpoint must never be cached at CloudFlare or in the browser.
  before_action :prevent_caching
  before_action :require_signed_in!

  # GET /user_list_state
  # Returns the signed-in user's lists + memberships scoped to Current.domain.
  # See docs/specs/user-lists-02a-add-to-list-widget.md for the response schema.
  def show
    domain = Current.domain
    subclass_names = list_subclasses_for(domain).map(&:name)

    lists = current_user.user_lists.where(type: subclass_names).order(:id).to_a

    items = if lists.any?
      UserListItem.where(user_list_id: lists.map(&:id))
        .pluck(:id, :user_list_id, :listable_type, :listable_id)
    else
      []
    end

    memberships = build_memberships(items)
    ensure_uid_cookie

    render json: {
      version: current_user.updated_at.to_i,
      domain: domain.to_s,
      user_id: current_user.id,
      lists: lists.map { |l| serialize_list(l) },
      memberships: memberships,
      # The cached HTML's <meta name="csrf-token"> belongs to whoever rendered the
      # cache (or no one). Issue a fresh per-session token here for client-side
      # mutations to send back via X-CSRF-Token. This endpoint is never cached.
      csrf_token: form_authenticity_token
    }
  end

  private

  # Maps the current request domain to the STI subclasses whose state should be loaded.
  # Books has no UserList subclass yet; movies/games each have one; music has two.
  def list_subclasses_for(domain)
    case domain
    when :music then [Music::Albums::UserList, Music::Songs::UserList]
    when :games then [Games::UserList]
    when :movies then [Movies::UserList]
    else []
    end
  end

  def serialize_list(list)
    icon_key = list.list_type.to_sym
    {
      id: list.id,
      type: list.class.name,
      list_type: list.list_type,
      name: list.name,
      default: list.default?,
      icon: list.class.list_type_icons[icon_key]
    }
  end

  # Backfills the tg_uid cookie for sessions established before the cookie was
  # introduced (or after a manual cookie-clear). Idempotent; setting the same
  # value is a no-op for the browser.
  def ensure_uid_cookie
    return if cookies[AuthController::TG_UID_COOKIE] == current_user.id.to_s
    cookies[AuthController::TG_UID_COOKIE] = {
      value: current_user.id.to_s,
      secure: Rails.env.production?,
      same_site: :lax
    }
  end

  # Shape: { listable_type => { listable_id (string) => [{list_id, item_id}, ...] } }
  # The item_id is needed so the modal can DELETE the underlying UserListItem
  # when a checkbox is unticked, without an extra round-trip to look it up.
  def build_memberships(items)
    items.each_with_object({}) do |(item_id, list_id, listable_type, listable_id), acc|
      acc[listable_type] ||= {}
      acc[listable_type][listable_id.to_s] ||= []
      acc[listable_type][listable_id.to_s] << {list_id: list_id, item_id: item_id}
    end
  end
end
